File: client/python/bone1/napkinUtil.py
####
# Copyright (c) 2013 Chris J Daly (github user cjdaly)
# All rights reserved. This program and the accompanying materials
# are made available under the terms of the Eclipse Public License v1.0
# which accompanies this distribution, and is available at
# http://www.eclipse.org/legal/epl-v10.html
#
# Contributors:
#   cjdaly - initial API and implementation
####

import urllib2, base64
import re

#
# Napkin config
DEVICE_ID="bone1"
NAPKIN_URLS=["http://192.168.2.78:4567", "http://192.168.2.35:4567", "http://192.168.2.236:4567"]
NAPKIN_URL_INDEX = 1

def getConfigUrl(napkinUrl):
	return napkinUrl + "/config"

def getConfigDeviceUrl(napkinUrl):
	return napkinUrl + "/config/" + DEVICE_ID

def getChatterUrl(napkinUrl):
	return napkinUrl + "/chatter"

#
# Napkin util

def getConfigValue(key):
    napkinConfigUrl=getConfigDeviceUrl(NAPKIN_URLS[NAPKIN_URL_INDEX]) + "?key=" + key
    configValue="??? error ???"
    try:
        request = urllib2.Request(napkinConfigUrl);
        authHeader = base64.encodestring('%s:%s' % (DEVICE_ID, DEVICE_ID)).replace('\n', '')
        request.add_header("Authorization", "Basic %s" % authHeader)
        configValue = urllib2.urlopen(request).read()
    except urllib2.URLError as urlError:
        print "URL Error in getConfigValue: {0}".format(urlError.strerror)
    except urllib2.HTTPError as httpError:
        print "HTTP Error in getConfigValue: {0}".format(httpError.strerror)
    #
    return configValue

def putConfigValue(key, value):
    napkinConfigUrl=getConfigDeviceUrl(NAPKIN_URLS[NAPKIN_URL_INDEX]) + "?key=" + key
    configValue="??? error ???"
    try:
	opener = urllib2.build_opener(urllib2.HTTPHandler)
        request = urllib2.Request(napkinConfigUrl, value);
        authHeader = base64.encodestring('%s:%s' % (DEVICE_ID, DEVICE_ID)).replace('\n', '')
        request.add_header("Authorization", "Basic %s" % authHeader)
        request.add_header("Content-Type", "text/plain")
	request.get_method = lambda: 'PUT'
        configValue = opener.open(request).read()
    except urllib2.URLError as urlError:
        print "URL Error in putConfigValue: {0}".format(urlError.strerror)
    except urllib2.HTTPError as httpError:
        print "HTTP Error in putConfigValue: {0}".format(httpError.strerror)
    #
    return configValue

def postDeviceConfig():
    postDeviceConfigUrl=getConfigUrl(NAPKIN_URLS[NAPKIN_URL_INDEX]) + "?sub=" + DEVICE_ID
    try:
        request = urllib2.Request(postDeviceConfigUrl, "")
        authHeader = base64.encodestring('%s:%s' % (DEVICE_ID, DEVICE_ID)).replace('\n', '')
        request.add_header("Authorization", "Basic %s" % authHeader)
        request.add_header("Content-Type", "text/plain")
        response = urllib2.urlopen(request)
        responseText = response.read()
    except urllib2.URLError as urlError:
        print "URL Error in postConfigNode: {0}".format(urlError.strerror)
    except urllib2.HTTPError as httpError:
        print "HTTP Error in postConfigNode: {0}".format(httpError.strerror)
    #
    return responseText

def getOrInitConfigValue(key, defaultValue):
    currentValue = getConfigValue(key)
    if (not currentValue):
        putConfigValue(key, defaultValue)
	currentValue = defaultValue
    return currentValue

def composeChatterText(cycle, startCount):
    chatterText = ""
    chatterText += "vitals.id=" + DEVICE_ID + "\n"
    chatterText += "vitals.startCount~i=" + str(startCount) + "\n"
    chatterText += "vitals.currentCycle~i=" + str(cycle) + "\n"
    return chatterText

def postChatter(chatterText):
    napkinChatterUrl=getChatterUrl(NAPKIN_URLS[NAPKIN_URL_INDEX]) + "?format=napkin_kv"
    responseText="??? error ???"
    try:
        request = urllib2.Request(napkinChatterUrl, chatterText)
        authHeader = base64.encodestring('%s:%s' % (DEVICE_ID, DEVICE_ID)).replace('\n', '')
        request.add_header("Authorization", "Basic %s" % authHeader)
        request.add_header("Content-Type", "text/plain")
        response = urllib2.urlopen(request)
        responseText = response.read()
    except urllib2.URLError as urlError:
        print "URL Error in postChatter: {0}".format(urlError.strerror)
    except urllib2.HTTPError as httpError:
        print "HTTP Error in postChatter: {0}".format(httpError.strerror)
    #
    return responseText

File: server/sinatra/napkin-util.rb
####
# Copyright (c) 2013 Chris J Daly (github user cjdaly)
# All rights reserved. This program and the accompanying materials
# are made available under the terms of the Eclipse Public License v1.0
# which accompanies this distribution, and is available at
# http://www.eclipse.org/legal/epl-v10.html
#
# Contributors:
#   cjdaly - initial API and implementation
####

module Napkin
  module Util
    module Conversion
      def parse_int(text)
        return nil if text.nil?
        begin
          return Integer(text)
        rescue ArgumentError => err
          return nil
        end
      end

      def parse_float(text)
        return nil if text.nil?
        begin
          return Float(text)
        rescue ArgumentError => err
          return nil
        end
      end
    end

    module Client
      require 'rest_client'
      include Conversion

      START_COUNT_KEY = "napkin.systems.start_count~i"

      #
      def get_chatter_url(device_id)
        return "http://#{device_id}:#{device_id}@localhost:4567/chatter"
      end

      def get_config_url(device_id)
        return "http://#{device_id}:#{device_id}@localhost:4567/config"
      end

      def increment_start_count(device_id)
        # create config area for device (if absent)
        config_url = get_config_url(device_id)
        RestClient.post(config_url, "", {:params => {'sub' => device_id}})

        # get current start_count
        device_config_url = "#{config_url}/#{device_id}"
        start_count_text = RestClient.get(device_config_url, {:params => {'key' => START_COUNT_KEY}})
        start_count = parse_int(start_count_text) || 0

        # increment and put value back to server config
        start_count += 1
        RestClient.put(device_config_url, start_count.to_s, {:params => {'key' => START_COUNT_KEY}})
        puts "#{device_id} starts: #{start_count}"
      end

      def chatter_sensor_data(data, device_id, key_prefix_filter)
        chatter_text = ""
        data.keys.each do |key|
          key_prefix_filter.each do |prefix|
            if (key.start_with?(prefix)) then
              chatter_text << "#{key} = #{data[key]}\n"
            end
          end
        end

        begin
          RestClient.post(get_chatter_url(device_id), chatter_text)
        rescue StandardError => err
          puts "Error: #{err}\n#{err.backtrace}"
        end
      end

      def process_sensor_data(sensor_uart, sensor_data)
        line = sensor_uart.gets
        if (!line.nil? && line.include?('=')) then
          key, value = line.split('=', 2)
          key.strip! ; value.strip!
          if (sensor_data.nil?) then
            if ((key == "state.vitalsAndSensorsUpdated") && (value == "false")) then
              sensor_data = {}
            end
          else
            if ((key == "state.vitalsAndSensorsUpdated") && (value == "true")) then
              if (sensor_data["state.vitalsAndSensorsUpdated"] == "false") then
                chatter_sensor_data(sensor_data, DEVICE_ID, CHATTER_KEY_PREFIXES)
              end
              sensor_data.clear
            else
              sensor_data[key] = value
            end
          end
        end
      end

    end

  end
end
